# 1. Beginner

In [1]:
import requests

In [2]:
response = requests.get("http://api.open-notify.org/this-api-doesnt-exist")

In [4]:
# response object
response

<Response [404]>

In [6]:
# using response.status code
response.status_code


404

### API Status Codes

1. 200: Everything went okay, and the result has been returned (if any).
2. 301: The server is redirecting you to a different endpoint. This can happen when a company switches domain names, or an endpoint name is changed.
3. 400: The server thinks you made a bad request. This can happen when you don’t send along the right data, among other things.
4. 401: The server thinks you’re not authenticated. Many APIs require login ccredentials, so this happens when you don’t send the right credentials to access an API.
5. 403: The resource you’re trying to access is forbidden: you don’t have the right permissions to see it.
6. 404: The resource you tried to access wasn’t found on the server.
7. 503: The server is not ready to handle the request.

**It is very important to read the documentation of the api**

[open api](http://open-notify.org/): this gives access to the data about the ISS(Internation Space Station)

**It is a very simple API and doesn't require authentication**

1. there can be multiple API's on a server and each of these API is called an endpoint

2. The documentation says that it requires no output

In [7]:
response = requests.get("http://api.open-notify.org/astros.json")

In [8]:
response.status_code

200

In [11]:
response.json()

{'message': 'success',
 'number': 7,
 'people': [{'craft': 'ISS', 'name': 'Sergey Ryzhikov'},
  {'craft': 'ISS', 'name': 'Kate Rubins'},
  {'craft': 'ISS', 'name': 'Sergey Kud-Sverchkov'},
  {'craft': 'ISS', 'name': 'Mike Hopkins'},
  {'craft': 'ISS', 'name': 'Victor Glover'},
  {'craft': 'ISS', 'name': 'Shannon Walker'},
  {'craft': 'ISS', 'name': 'Soichi Noguchi'}]}

In [16]:
import json

#json.dumps converts a python object like dict to a string type
str_obj = json.dumps(response.json())
print(type(str_obj))

#json.loads coverts a json string to a python object
python_obj = json.loads(str_obj)
print(type(python_obj))

<class 'str'>
<class 'dict'>


## Using an API with Query Parameters

In [17]:
parameters = {'lat': 38.907192, 'lon': -77.036873}
response = requests.get("http://api.open-notify.org/iss-pass.json", params=parameters)

In [18]:
response

<Response [200]>

In [21]:
pass_times = response.json()['response']

In [22]:
risetimes = []

for d in pass_times:
    time = d['risetime']
    risetimes.append(time)

print(risetimes)

[1607118811, 1607124622, 1607130432, 1607184792]


In [23]:
from datetime import datetime

times = []

for rt in risetimes:
    time = datetime.fromtimestamp(rt)
    times.append(time)
    print(time)

2020-12-04 16:53:31
2020-12-04 18:30:22
2020-12-04 20:07:12
2020-12-05 11:13:12


# 2. Intermediate

**Getting Music Data with the Last.fm API using Python**

We'll learn about

1. How to authenticate yourself with an API

2. working within the guidelines of the API ( rate limiting:# of requests per second)

3. Using Pagination to work with large reponses

**Objective:We’ll be building a dataset of popular artists using their API.**

### Step 1: Get yourself autheticated 
https://www.dataquest.io/blog/last-fm-api-python/

Application name	Learning API

API key	b2d2cad4467494beed43f96a7a60330c

Shared secret	70f4c9e8ab7539d85840d3246ca074e8

Registered to	runiyal


import requests

headers = {
    'user_agent': user_agent
}

parameters = {
    'api_key': api_key,
    'method': 'chart.gettopartists',
    'format': 'json'
}

r = requests.get('http://ws.audioscrobbler.com/2.0/', params=parameters, headers=headers)


In [31]:
user_agent='runiyal'
api_key = 'b2d2cad4467494beed43f96a7a60330c'

In [32]:
def lastfm_get(parameters):
    
    headers = {
        'user_agent': user_agent
    } 
    
    parameters['api_key'] = api_key
    parameters['format'] = 'json'
    
    url = 'http://ws.audioscrobbler.com/2.0/'
    
    return(requests.get(url, params=parameters, headers=headers))

In [35]:
params = {
    'method': 'chart.gettopartists'
}
r =lastfm_get(params)

In [40]:
import json

def jprint(obj):
    # create a formatted string of the Python JSON object
    text = json.dumps(obj, sort_keys=True, indent=4)
    print(text)

jprint(r.json())

{
    "artists": {
        "@attr": {
            "page": "1",
            "perPage": "50",
            "total": "3779091",
            "totalPages": "75582"
        },
        "artist": [
            {
                "image": [
                    {
                        "#text": "https://lastfm.freetls.fastly.net/i/u/34s/2a96cbd8b46e442fc41c2b86b821562f.png",
                        "size": "small"
                    },
                    {
                        "#text": "https://lastfm.freetls.fastly.net/i/u/64s/2a96cbd8b46e442fc41c2b86b821562f.png",
                        "size": "medium"
                    },
                    {
                        "#text": "https://lastfm.freetls.fastly.net/i/u/174s/2a96cbd8b46e442fc41c2b86b821562f.png",
                        "size": "large"
                    },
                    {
                        "#text": "https://lastfm.freetls.fastly.net/i/u/300x300/2a96cbd8b46e442fc41c2b86b821562f.png",
                        "si

In [41]:
jprint(r.json()['artists']['@attr'])


{
    "page": "1",
    "perPage": "50",
    "total": "3779091",
    "totalPages": "75582"
}


In [46]:
import requests
import json


user_agent='runiyal'
api_key = 'b2d2cad4467494beed43f96a7a60330c'

def lastfm_get(params):
    
    params['format'] = 'json'
    params['api_key'] = api_key
    
    headers = {
        'user_agent': user_agent
    }
    
    url = 'http://ws.audioscrobbler.com/2.0/'
    return(requests.get(url, params=params, headers=headers))

In [50]:
lastfm_get({'method':'chart.gettopartists'}).json()['artists']['@attr']

{'page': '1', 'perPage': '50', 'totalPages': '75582', 'total': '3779091'}

In [54]:
import requests_cache

requests_cache.install_cache()

In [59]:
import time
from IPython.core.display import clear_output

responses = []

page=1
total_pages=9999

while page<=total_pages:
    parameters = {
                'method':'chart.gettopartists',
                 'limit':500,
                 'page':page
    }
    
    r = lastfm_get(parameters)
    total_pages = int(r.json()['artists']['@attr']['totalPages'])
    page = int(r.json()['artists']['@attr']['page'])
    print("Requesting page {}/{}".format(page, total_pages))
    
    clear_output(wait=True)
    
    if r.status_code != 200:
        print(r.text())
        break
        
    responses.append(r)
        
        
    # if it's not a cached result, sleep
    if not getattr(response, 'from_cache', False):
        time.sleep(0.25)
        
    page += 1

Requesting page 7575/7575


### Processing the Data

In [65]:
import pandas as pd

In [69]:
frames = [pd.DataFrame(r.json()['artists']['artist']) for r in responses]
combined_df = pd.concat(frames)

In [72]:
artists = combined_df.drop('image', axis=1)
artists.head()

name  playcount listeners                                  mbid  \
0  Ariana Grande  197537360   1491590  f4fdbb4c-e4b7-47a0-b83b-d91bbfcfa387   
1     The Weeknd  142832265   1738035  c8b03190-306c-4120-bb0b-6f2ebfc06ea9   
2  Billie Eilish   78427275    976658                                         
3     Kanye West  322825815   4735662  164f0d73-1234-4e2c-8743-d77bf2191051   
4    Miley Cyrus   93842111   2361097  7e9bd05a-117f-4cce-87bc-e011527a8b18   

                                       url streamable  
0  https://www.last.fm/music/Ariana+Grande          0  
1     https://www.last.fm/music/The+Weeknd          0  
2  https://www.last.fm/music/Billie+Eilish          0  
3     https://www.last.fm/music/Kanye+West          0  
4    https://www.last.fm/music/Miley+Cyrus          0